File: packages/expo-dual-camera/ios/ExpoDualCameraModule.swift
import AVFoundation
import ExpoModulesCore

/// JS-facing module: thin delegate to `DualCameraCaptureController` with exactly-once Promise settlement.
public class ExpoDualCameraModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoDualCamera")

    AsyncFunction("isSupported") { () -> Bool in
      DualCameraTurboSafe.isMultiCamSupported()
    }

    AsyncFunction("getCameraPermissionsAsync") { () -> [String: Any] in
      DualCameraCaptureController.permissionResponse()
    }

    AsyncFunction("requestCameraPermissionsAsync") { (promise: Promise) in
      let settled = DualCameraPromiseGuard(promise)
      DualCameraTurboSafe.invokeAsync(guard: settled) { settled in
        if AVCaptureDevice.authorizationStatus(for: .video) == .authorized {
          settled.resolve(DualCameraCaptureController.permissionResponse())
          return
        }
        AVCaptureDevice.requestAccess(for: .video) { _ in
          var innerNs: NSError?
          let innerOk = ObjcPerformCatching({
            settled.resolve(DualCameraCaptureController.permissionResponse())
          }, &innerNs)
          if !innerOk {
            settled.reject(
              "E_DUAL_CAMERA_NS_EXCEPTION",
              (innerNs as NSError?)?.localizedDescription ?? "NSException"
            )
          }
        }
      }
    }

    AsyncFunction("takePictureAsync") { (side: String, options: [String: Any]?, promise: Promise) in
      let settled = DualCameraPromiseGuard(promise)
      DualCameraTurboSafe.invokeAsync(guard: settled) { settled in
        let opts = CaptureOptions(from: options)
        DualCameraCaptureController.shared.takePicture(side: side, options: opts) { result in
          switch result {
          case .success(let data):
            settled.resolve(data)
          case .failure(let error):
            settled.reject("E_CAPTURE", error.localizedDescription)
          }
        }
      }
    }

    Function("pausePreview") {
      DualCameraTurboSafe.invokeSync {
        DualCameraCaptureController.shared.pausePreview()
      }
    }

    Function("resumePreview") {
      DualCameraTurboSafe.invokeSync {
        DualCameraCaptureController.shared.resumePreview()
      }
    }

    Function("swapRecordingLayout") {
      DualCameraTurboSafe.invokeSync {
        DualCameraCaptureController.shared.swapRecordingLayout()
      }
    }

    AsyncFunction("concatVideoSegments") { (options: [String: Any]?, promise: Promise) in
      let settled = DualCameraPromiseGuard(promise)
      DualCameraTurboSafe.invokeAsync(guard: settled) { settled in
        let rawUris = options?["uris"] as? [String] ?? []
        let urls = rawUris.compactMap { URL(string: $0) }
        guard !urls.isEmpty else {
          settled.reject("E_CONCAT", "No video segments to merge")
          return
        }
        DualCameraVideoConcat.concat(urls: urls) { result in
          switch result {
          case .success(let data):
            settled.resolve(data)
          case .failure(let error):
            settled.reject("E_CONCAT", error.localizedDescription)
          }
        }
      }
    }

    AsyncFunction("startRecording") { (options: [String: Any]?, promise: Promise) in
      guard let settled = DualCameraTurboSafe.acquireStartRecordingPromise(promise) else {
        DualCameraPromiseGuard(promise).reject(
          "E_RECORDING_BUSY",
          "startRecording already in progress"
        )
        return
      }
      DualCameraTurboSafe.invokeAsync(guard: settled) { settled in
        DualCameraCaptureController.shared.startRecording(options: options) { result in
          switch result {
          case .success:
            settled.resolve(true)
          case .failure(let error):
            settled.reject("E_RECORDING_START", error.localizedDescription)
          }
        }
      }
    }

    AsyncFunction("stopRecording") { (promise: Promise) in
      guard let settled = DualCameraTurboSafe.acquireStopRecordingPromise(promise) else {
        DualCameraPromiseGuard(promise).reject(
          "E_RECORDING_BUSY",
          "stopRecording already in progress"
        )
        return
      }
      DualCameraTurboSafe.invokeAsync(guard: settled) { settled in
        DualCameraCaptureController.shared.stopRecording { result in
          switch result {
          case .success(let data):
            settled.resolve(data)
          case .failure(let error):
            settled.reject("E_RECORDING_STOP", error.localizedDescription)
          }
        }
      }
    }

    View(DualCameraView.self) {
      Events("onCameraReady", "onMountError")

      Prop("side") { (view: DualCameraView, side: String) in
        view.setSide(side)
      }
      Prop("lens") { (view: DualCameraView, lens: String) in
        view.setLens(lens)
      }
      Prop("zoom") { (view: DualCameraView, zoom: Double) in
        view.setZoom(zoom)
      }
      Prop("enableTorch") { (view: DualCameraView, enabled: Bool) in
        view.setEnableTorch(enabled)
      }
      Prop("flash") { (view: DualCameraView, mode: String) in
        view.setFlash(mode)
      }
      Prop("mirror") { (view: DualCameraView, mirror: Bool) in
        view.setMirror(mirror)
      }
      Prop("autofocus") { (view: DualCameraView, mode: String) in
        view.setAutofocus(mode)
      }
    }
  }
}

File: packages/expo-dual-camera/android/src/main/java/expo/modules/dualcamera/ExpoDualCameraModule.kt
package expo.modules.dualcamera

import android.Manifest
import android.content.pm.PackageManager
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.content.ContextCompat
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class ExpoDualCameraModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoDualCamera")

    // MARK: - Support Check

    AsyncFunction("isSupported") { promise: Promise ->
      val context = appContext.reactContext ?: run {
        promise.resolve(false); return@AsyncFunction
      }
      val future = ProcessCameraProvider.getInstance(context)
      future.addListener({
        val provider = future.get()
        val supported = provider.availableConcurrentCameraInfos.isNotEmpty()
        promise.resolve(supported)
      }, ContextCompat.getMainExecutor(context))
    }

    // MARK: - Permissions

    AsyncFunction("getCameraPermissionsAsync") { promise: Promise ->
      val context = appContext.reactContext ?: run {
        promise.reject("E_CONTEXT", "Context not available", null); return@AsyncFunction
      }
      val granted = ContextCompat.checkSelfPermission(
        context, Manifest.permission.CAMERA
      ) == PackageManager.PERMISSION_GRANTED
      promise.resolve(permissionResponse(granted, canAskAgain = true))
    }

    AsyncFunction("requestCameraPermissionsAsync") { promise: Promise ->
      val context = appContext.reactContext ?: run {
        promise.reject("E_CONTEXT", "Context not available", null); return@AsyncFunction
      }
      val alreadyGranted = ContextCompat.checkSelfPermission(
        context, Manifest.permission.CAMERA
      ) == PackageManager.PERMISSION_GRANTED

      if (alreadyGranted) {
        promise.resolve(permissionResponse(true, canAskAgain = true))
        return@AsyncFunction
      }

      val permissions = appContext.permissions
      if (permissions == null) {
        promise.reject("E_PERMISSIONS", "Permissions module not available", null)
        return@AsyncFunction
      }

      permissions.askForPermissionsWithPermissionsManager(
        promise,
        Manifest.permission.CAMERA
      )
    }

    // MARK: - Photo Capture

    AsyncFunction("takePictureAsync") { side: String, options: Map<String, Any>?, promise: Promise ->
      val context = appContext.reactContext ?: run {
        promise.reject("E_CONTEXT", "Context not available", null); return@AsyncFunction
      }
      DualCameraSessionManager.takePicture(side, options, context) { result ->
        result.onSuccess { data -> promise.resolve(data) }
        result.onFailure { error ->
          promise.reject("E_CAPTURE", error.message ?: "Capture failed", error)
        }
      }
    }

    // MARK: - Session Control

    Function("pausePreview") {
      DualCameraSessionManager.pausePreview()
    }

    Function("resumePreview") {
      DualCameraSessionManager.resumePreview()
    }

    // MARK: - View

    View(DualCameraView::class) {
      Events("onCameraReady", "onMountError")

      Prop("side") { view: DualCameraView, side: String ->
        view.setSide(side)
      }
      Prop("lens") { view: DualCameraView, lens: String ->
        view.setLens(lens)
      }
      Prop("zoom") { view: DualCameraView, zoom: Double ->
        view.setZoom(zoom)
      }
      Prop("enableTorch") { view: DualCameraView, enabled: Boolean ->
        view.setEnableTorch(enabled)
      }
      Prop("flash") { view: DualCameraView, mode: String ->
        view.setFlash(mode)
      }
      Prop("mirror") { view: DualCameraView, mirror: Boolean ->
        view.setMirror(mirror)
      }
      Prop("autofocus") { view: DualCameraView, mode: String ->
        view.setAutofocus(mode)
      }
    }
  }

  private fun permissionResponse(granted: Boolean, canAskAgain: Boolean): Map<String, Any> {
    return mapOf(
      "status" to if (granted) "granted" else "denied",
      "granted" to granted,
      "canAskAgain" to canAskAgain,
      "expires" to "never"
    )
  }
}
